Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,175 @@ lemma iff_flat_and_lTensor_reflects_triviality :
 
 end faithful
 
+section complex
+
+private lemma range_le_ker_of_exact_rTensor [fl : FaithfullyFlat R M]
+    ⦃N1 N2 N3 : Type*⦄
+    [AddCommGroup N1] [Module R N1]
+    [AddCommGroup N2] [Module R N2]
+    [AddCommGroup N3] [Module R N3]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+    (ex : Function.Exact (l12.rTensor M) (l23.rTensor M)) :
+    LinearMap.range l12 ≤ LinearMap.ker l23 := by
+  intro x hx
+  simp only [LinearMap.mem_ker]
+  obtain ⟨y, rfl⟩ := hx
+  have : ∀ (n1 : N1) (m : M), l23 (l12 n1) ⊗ₜ[R] m = 0 := fun n1 m ↦
+    ex.apply_apply_eq_zero (n1 ⊗ₜ[R] m)
+  have eq1 := this y
+  by_contra! hxx
+  let E : Submodule R N3 := Submodule.span R {l23 (l12 y)}
+  have hE : Nontrivial E :=
+    ⟨0, ⟨⟨l23 (l12 y), Submodule.mem_span_singleton_self _⟩, Subtype.coe_ne_coe.1 hxx.symm⟩⟩
+  have eq0: (⊤ : Submodule R (E ⊗[R] M)) = ⊥ := by
+    ext x
+    simp only [Submodule.mem_top, Submodule.mem_bot, true_iff]
+    have mem : x ∈ (⊤ : Submodule R _) := ⟨⟩
+    rw [← TensorProduct.span_tmul_eq_top, mem_span_set] at mem
+    obtain ⟨c, hc, rfl⟩ := mem
+    choose b a hy using hc
+    let r :  ⦃a : E ⊗[R] M⦄ → a ∈ ↑c.support → R := fun a ha =>
+      Submodule.mem_span_singleton.1 (b ha).2 |>.choose
+    have hr : ∀ ⦃i : E ⊗[R] M⦄ (hi : i ∈ c.support), b hi =
+        r hi • ⟨l23 (l12 y), Submodule.mem_span_singleton_self _⟩ := by
+      intro i hi
+      ext
+      exact Submodule.mem_span_singleton.1 (b hi).2 |>.choose_spec.symm
+
```

Comment:
I think this empty line is here to signify a "break" in the proof. Please use some comments to guide the reader a bit.

Reply:
I have added some comments about the proof